### DISCUSSION TITLE
Topics for documentation

### DISCUSSION THREAD

**Discussion by MickL** [nitrojs/nitro#2753] [Ideas]
I may update the list by time. Feel free to add your ideas!

* Everything might be more beginner friendly, guide like. E.g. see AdonisJS docs for a good ref: Start with Installation, configuration/env, folder structure, how to create controllers (explains -> file based routing + event handler), deployment
* Provide best practices. For example in NestJS I would spread a lot of logic into services. E.g. having auth routes (controller) that use auth service, password hashing service, user service. In Nuxt and Nitro I feel like I need to do everything in the controller (event handler) or I write helper functions (utils or composables) but they are located in way different folders because of the file based routing which bothers me a bit especially when the project grows. What is a best practice to manage more complex projects?
* Body and Queryparams validation (this is not even documented in H3, just in a blog post)
    * Personally i would use ~~Zod~~ Valibot as an example here (Valibot is tree shakable which Zod is more and seems more robust because of its structure)
* Logging
   * Is not documented at all. I guess consola is already integrated and should be used? If yes: Add info that this is the case and how to use it.
* Usage of databases
    * Personally I would use Drizzle here and create a simple use-db.ts file
* i18n
   * Maybe how to translate an error message (same concept will be useful for stuff like generating PDFs, sending emails, etc.)
* Authentication and sessions
    * Login route with username/email and password, check password, create session
    * Common pitfall: When saving passwords into database, hashing the password is not possible on edge workers (Cloudflare has functions but no good alternative). What is a possible workaround or best practice?
    * First I thought this could use Lucia as an example. But I am pretty sure H3 can already handle sessions and there is no need for Lucia. If this is true this information is crucial because probably "everyone" will use Lucia even tho there is no need for it when using H3 / Nitro / Nuxt. I didnt try (also because session docs are missing) but from my imagination only thing needs to be done is save sessions into database so they can expire.
* Testing should be a huge part explaining how to test functions separate and routes e2e, how to mock things and show best practices, see also #2303
* How to develop third party libraries / modules / plugins
* I would also grab everything from the Nuxt docs and place it in the Nitro docs and then remove it from the Nuxt docs and link to Nitro docs because the server part can only be incomplete if it is not a full duplicate of the Nitro docs. https://nuxt-com.translate.goog/docs/guide/directory-structure/server
* Provide benchmarks, maybe in comparison also with other Frameworks (Express, Fastify, NestJS, AdonisJS, Bun) - #2534
* How to accept files and then save them: `readMultipartFormData` then `useStorage().setItemRaw`
* For each code-example I would add imports (NestJS and AdonisJS do this too) and NOT leave the out because of auto-imports
* How to get route handler `event` within a util function. Should be `useEvent()` but has limitations and requires experimental flag.

**Comment by MickL** [comment]
Added logger, currently Nitro docs contain no info about logging

**Reply by pi0** [reply]
(there is no built-in feature today, we might include a how to add consola recepie though)

**Reply by MickL** [reply]
Yes, I think a lot of features can be resolved just by adding a guide / recipe to the docs. Logging is a good example.

**Comment by Exotelis** [comment]
I'd like to suggest layers. Please see [2784](https://github.com/unjs/nitro/discussions/2784) for details

**Comment by lizyChy0329** [comment]
The api's for their dependency libraries are nowhere to be found in the documentation. nitro/nuxt/vite all have the same problem.

**Comment by MickL** [comment]
Added "How to accept files and then save them". This is super simple but took me forever to find out because there is barely until information how to get the file from the body, answer: `readMultipartFormData`. Then to save the file `useStorage().setItemRaw()` needs to be used instead if `useStorage().setItem()`, this wasnt trivial for me too.

Also when looking at unjs/unstorage it seems like we need to import and create a driver object, e.g. `driver: fsDriver({ base: "./tmp" }),` but in Nitro it is just needed to add the name (super simple but not trivial as well): `driver: 'fs'`
